Vmd pch rootbus v3

sdurawa opened the pull request:
New version with applied Bjorn remarks

Commit:
PCI: vmd: Add vmd_configure_cfgbar()

Move the VMD CFGBAR initialization code to a new helper. No
functional changes.

Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>
Reviewed-by: Mariusz Tkaczyk <mariusz.tkaczyk@linux.intel.com>
Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>

Commit:
PCI: vmd: Convert bus and busn_start to an array

Bus and busn_start are converted from scalar to an array to support
multiple VMD buses in the future. No functional changes.

Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>
Reviewed-by: Mariusz Tkaczyk <mariusz.tkaczyk@linux.intel.com>
Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>

mtkaczyk reviewed (commented):
Patch1 :
`PCI: vmd: Create bus enumeration helper`

Please change to:
`PCI: vmd: Add vmd_bus_enumeration()`

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -790,6 +790,48 @@ static void vmd_configure_cfgbar(struct vmd_dev *vmd)
+		flags &= ~IORESOURCE_MEM_64;
+
+	vmd->resources[resource_number] = (struct resource){
+		.name = kasprintf(GFP_KERNEL, "VMD MEMBAR%d",
```
It is no longer const,  shouldn't it be freed in vmd_remove?

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -855,26 +861,28 @@ static int vmd_create_bus(struct vmd_dev *vmd, struct pci_sysdata *sd,
 
-	vmd_copy_host_bridge_flags(pci_find_host_bridge(vmd->dev->bus),
-				   to_pci_host_bridge(vmd->bus->bridge));
+	vmd_copy_host_bridge_flags(
```
formatting

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -536,6 +565,10 @@ static void vmd_domain_reset(struct vmd_dev *vmd)
 
+	if (vmd->bus1_rootbus)
+		max_buses +=
+			resource_size(&vmd->resources[VMD_RES_BUS1_CFGBAR]) + 2;
```
formatting

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -679,6 +724,19 @@ static int vmd_get_bus_number_start(struct vmd_dev *vmd)
+				return -ENODEV;
+			}
+
+			/* BUS0 (IOC) start number */
```
Bjorn doesn't like our IOC, PCH nomenclature, please challenge that :)

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -679,6 +724,19 @@ static int vmd_get_bus_number_start(struct vmd_dev *vmd)
+
+			/* BUS0 (IOC) start number */
+			vmd->busn_start[VMD_BUS_0] = VMD_RESTRICT_2_BUS_START;
+			/* BUS1 (PCH) start number */
```
as above (PCH)

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -846,44 +929,84 @@ static void vmd_configure_membar(struct vmd_dev *vmd,
-	vmd->bus[VMD_BUS_0] = pci_create_root_bus(&vmd->dev->dev,
-						  vmd->busn_start[VMD_BUS_0],
+	vmd->bus[bus_number] = pci_create_root_bus(&vmd->dev->dev,
+						  vmd->busn_start[bus_number],
```
formatting

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -846,44 +929,84 @@ static void vmd_configure_membar(struct vmd_dev *vmd,
-	     "Can't create symlink to domain\n");
+	WARN(sysfs_create_link(
+		     &vmd->dev->dev.kobj, &vmd->bus[bus_number]->dev.kobj,
+		     kasprintf(GFP_KERNEL, "domain bus%d", bus_number)),
```
formatting here and line above

mtkaczyk replied:
You can't just change link. "domain" link is there and it must stay as is. For example, you are breaking this code:
https://github.com/md-raid-utilities/mdadm/blob/main/platform-intel.c#L126

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -421,8 +421,22 @@ static void vmd_remove_irq_domain(struct vmd_dev *vmd)
+	u32 offset;
+
+	/*
+	 * VMD WA: for PCH rootbus, bus number is set to VMD_PRIMARY_BUS1
```
Please explain WA

mtkaczyk reviewed (changes requested):
Bjorn asked what is happening if this workaround is not applied, please elaborate :)

Commit:
PCI: vmd: Add workaround for rootbus number hardwired to fixed non-zero value

VMD BUS1 rootbus primary number is 0x80 and pci_scan_bridge_extend() cannot
assign it as "hard-wired to 0" and marks setup as broken. To avoid falling
into bridge configuration invalid condition, BUS1 number has to be the same
as BUS1 primary number.

Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>
Reviewed-by: Mariusz Tkaczyk <mariusz.tkaczyk@linux.intel.com>
Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>

Commit:
PCI: vmd: Add vmd_create_bus()

Move the VMD bus initialization code to a new helper. No
functional changes.

Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>
Reviewed-by: Mariusz Tkaczyk <mariusz.tkaczyk@linux.intel.com>
Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>

Commit:
PCI: vmd: Replace hardcoded values with enum and defines

Add enum vmd_resource type to replace hardcoded values. Add defines for vmd bus
start number based on VMD restriction value. No functional changes.

Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>
Reviewed-by: Mariusz Tkaczyk <mariusz.tkaczyk@linux.intel.com>
Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>

Commit:
PCI: vmd: Add vmd_configure_membar() and vmd_configure_membar1_membar2()

Move the MEMBAR1 and MEMBAR2 registry initialization code to a new helper
functions. No functional changes.

Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>
Reviewed-by: Mariusz Tkaczyk <mariusz.tkaczyk@linux.intel.com>
Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>

Commit:
PCI: vmd: Add support for second rootbus under VMD

Starting from Intel Arrow Lake VMD enhacement introduces second rootbus support
with fixed root bus number (0x80). It means that all 3 MMIO BARs exposed by
VMD are shared now between both buses (current BUS0 and new BUS1).
This patch add new BUS1 enumeration and divide MMIO space to be shared between
both rootbuses. Due to enumeration issues with rootbus hardwired to a
fixed non-zero value, this patch will work with a workaround proposed in next
patch. Without workaraund user won't see attached devices behind BUS1 rootbus.

Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>
Reviewed-by: Mariusz Tkaczyk <mariusz.tkaczyk@linux.intel.com>
Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>

Commit:
PCI: vmd: Add vmd_bus_enumeration()

Move the vmd bus enumeration code to a new helper. No
functional changes.

Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>
Reviewed-by: Mariusz Tkaczyk <mariusz.tkaczyk@linux.intel.com>
Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>
